src/main/scala/leon/purescala/Constructors.scala

```diff
@@ -144,7 +145,7 @@ object Constructors {
 
     resType match {
       case Some(tpe) =>
-        casesFiltered.filter(c => isSubtypeOf(c.rhs.getType, tpe) || isSubtypeOf(tpe, c.rhs.getType))
+        casesFiltered.filter(c => isSubtypeOf(c.rhs.getType, tpe) || isSubtypeOf(tpe, c.rhs.getType) || c.optGuard.nonEmpty)
```

Comment:
Indeed, it was probably first introduced as a pseudo-fix to the issue with swapping guard and rhs.
Thanks for noticing. I removed it.